### PR TITLE
Support base tags for interface items

### PR DESCRIPTION
**Pull request by wthrowe**
Unfortunately it is not possible to make this work automatically,
because there is no way to determine the base classes of a class.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files) or `tests/Unit/TestingFramework.hpp` (only in tests)
  2. Blank line (only in cpp files and tests)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

**Review by kidder — DISMISSED** [review]
@nilsdeppe this is ready for you to look at

File: src/Domain/Tags.hpp
Comment (nilsdeppe) [inline src/Domain/Tags.hpp]: I'm not clear on how these can be a base tag. If it's a base tag it must inherit off of `db::BaseTag` and not `db::SimpleTag` or `db::ComputeTag`. To me it looks like it's currently confused between being a base tag and a more concrete tag. Could you add a comment explaining how these aren't violating the base tag requirements?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Domain/Tags.hpp
Comment (wthrowe) [reply src/Domain/Tags.hpp]: I meant it in the sense of a base class (and also the "base tag mechanism" that it is implementing).  I am not aware of a term for the part of the base tag mechanism that isn't related to `db::BaseTag`.  It is actually not possible to use a `db::BaseTag`-style base tag with interface items because interface items require `db::item_type` to work.

Comment (nilsdeppe) [reply src/Domain/Tags.hpp]: Ah, I understand, sorry. Yes, we should try and find a way to differentiate those :( I'm currently not coming up with any name that isn't horrible and super vague...

Could the `InterfaceBaseTag` then always be a simple tag? The design goal was that you could have compute tags inherit off of simple tags, in which case the return type lookup is more efficient at compile time. Or is there something that currently relies of the base class tag also being marked as a compute tag? I now that some parts of the DataBox code, specifically item_type, avoids looking up derived tags if the tag is not a `db::BaseTag` and I'm worried that this might cause unexpected and confusing errors in the future.

Comment (wthrowe) [reply src/Domain/Tags.hpp]: I can't think of any realistic cases where this would try to use a compute tag as a base class.  You could confuse it if you tried, but I'm not coming up with anything that would arise naturally.

Most of this mess if from avoiding compiler warnings from inheriting from the tag markers twice.  I think I'll just give up and mark them as virtual.

File: src/Domain/Tags.hpp
Comment (nilsdeppe) [inline src/Domain/Tags.hpp]: Sorry, I'm obviously still confused on something. I'm going to say what I think is happening, and then please correct me where I'm wrong:

- The empty case, line 260, is for when there is no base class and so there is base tag mechanism active.
- The second case, line 263, is used if the tag as a `base` type alias, and is used to make `InterfaceImpl` inherit off of `Tag::base`. This means that `InterfaceBaseTag` is never used as a base tag, but is purely a method of getting the correct base class onto the interface tag.

If `InterfaceBaseTag` is only used for generating the inheritance maybe it should be named something like `GetBaseTagIfPresent`? I think with that sort of a name change this code would be easier to understand, since initially I was expecting the class itself to be used as a tag.

One small note:
The one thing I'm not sure on is if this will work with a `db::BaseTag`. Do you want it to? If not, it would be useful to state that, if so, it should be tested.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash things :)

File: src/Domain/Tags.hpp
Comment (wthrowe) [reply src/Domain/Tags.hpp]: Your understanding of what is going on is correct, and I like your renaming suggestion.

It does not currently work with `db::BaseTag` and I was going to say that there was a good reason for that, but now I think I know how to fix it.

**Comment by kidder** [comment]
squash, then ping @nilsdeppe

File: tests/Unit/Domain/Test_InterfaceItems.cpp
Comment (nilsdeppe) [inline tests/Unit/Domain/Test_InterfaceItems.cpp]: Could you check that having a compute item that has `ComputeBase` as its argument works correctly?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash this.

I'm traveling to Germany tomorrow (Saturday) for a family emergency, so if I can't approve this, please just dismiss my review and merge it :) @kidder you should be able to dismiss it

File: tests/Unit/Domain/Test_InterfaceItems.cpp
Comment (wthrowe) [reply tests/Unit/Domain/Test_InterfaceItems.cpp]: Doesn't work, and I can't think of a way to fix it.  I may have to drop `db::BaseTag` support again, but I'll think about it for a few days first.

Comment (nilsdeppe) [reply tests/Unit/Domain/Test_InterfaceItems.cpp]: Okay, well I'll leave it to you to decide whether base tags in this context are important enough to be worth putting in the effort (now, or maybe later?) :)

**Comment by kidder** [comment]
@nilsdeppe please look, I am happy for @wthrowe to squash

File: src/Domain/Tags.hpp
Comment (nilsdeppe) [inline src/Domain/Tags.hpp:307]: This is not related to the PR (I thought of this will looking at the tags in the tests), but should the name here not include the tag being wrapped or something? Whenever that gets changed (an issue and a later PR is totally fine with me) we should make sure to have a test for it too :)

**Review by nilsdeppe — DISMISSED** [review]
I'm happy for you to squash this

File: src/Domain/Tags.hpp
Comment (wthrowe) [reply src/Domain/Tags.hpp:307]: This is true of every prefix tag in the code.  Someone should just fix them all instead of filing a bug for each one individually whenever someone happens to think of it.

**Review by kidder — APPROVED** [review]

**Review by mar-celine — APPROVED** [review]

**Pull request merged**
